Adds the `buoy_detect` module

pranaypratyush opened the pull request:
- msgs for the above module
- adds topics in the `resources` stack

pranaypratyush committed:
Adds the 'buoy_detect' module
+ msgs for the above module
+ adds topics in the resources stack

icyflame commented:
@pranaypratyush 

1. Remove `vision_stack/buoy_detect/LICENSE.md` and `Makefile` (will be generated automatically by Cmake)
2. Has this been tested in the pool?

pranaypratyush commented:
Donno how that `Makefile` slipped in. Amending now.

prudhvid commented on vision_stack/buoy_detect/region_growing/src/grow.cpp:
```diff
@@ -0,0 +1,290 @@
+        and input pixel to deal overexposed regions of buoy.
+        This function implements CIEDE2000 algorithm to find the distance in color
+ */
+bool colorDistance(Vec3b a, Vec3b b, int colorThreshold, int whiteThreshold)
```
pass objects by const reference
const Vec3b& a, const Vec3b &b

pranaypratyush replied:
glad to hear from u.
@Nightfury18  your code.

prudhvid replied:
:wink:

prudhvid commented on vision_stack/buoy_detect/region_growing/src/grow.cpp:
```diff
@@ -0,0 +1,290 @@
+    Vec3b &colorC = input.at<Vec3b>(x, y);
+
+    //Yellow Pixel
+    if (colorflag == 2)
```
If possible(time permits), define these flags as const before and use those const variables

pranaypratyush replied:
@Nightfury18 take a look.

icyflame replied:
@prudhvid That is certainly required. A year down the line, we are going to find ourselves tracking the code to figure out what the integers really mean.

prudhvid commented on vision_stack/buoy_detect/region_growing/src/grow.cpp:
```diff
@@ -0,0 +1,290 @@
+    stringstream ss;
+
+    ss << x;
+    s += ss.str() + "-";
```
This might give strange values in other parts of code when x<0. 
I don't know if that is possible but I suggest to change the delimiting character to something like $

prudhvid commented on vision_stack/buoy_detect/region_growing/src/grow.cpp:
```diff
@@ -0,0 +1,290 @@
+        s = queue.front();
+        queue.pop_front();
+
+        std::size_t pos = s.find('-');
```
The previous comment at line 81 referes to this. e.g. when x<0 then pos = 0. hence put delimiting character as $

pranaypratyush committed:
Minor bug fix
 (still getting seg_fault though)

prudhvid commented on vision_stack/buoy_detect/region_growing/src/grow.cpp:
```diff
@@ -0,0 +1,290 @@
+    Vec3b seed = input.at<Vec3b>(sX, sY);
+    vector < vector<bool> > reach(input.rows, vector<bool>(input.cols, false));
+
+    list <string> queue;
```
Instead of all the stringToInt and IntToString conversion, you could do a simple

``` C++
list<pair<int, int> > posQueue; //renamed it because queue is a standard STL template
```

prudhvid commented on vision_stack/buoy_detect/region_growing/src/grow.cpp:
```diff
@@ -0,0 +1,290 @@
+    modifyPixel(filled, sX, sY, colorflag);
+
+    s = intToString(sX, sY);
+    queue.push_back(s);
```
This becomes

``` C++
posQueue.push_back(make_pair(sX, sY));
```

prudhvid commented on vision_stack/buoy_detect/region_growing/src/grow.cpp:
```diff
@@ -0,0 +1,290 @@
+        stringstream ssx(s.substr(0, pos));
+        stringstream ssy(s.substr(pos + 1));
+        ssx >> x;
+        ssy >> y;
```
These become 

```
x = posQueue.front().first;
y = posQueue.front().second;
```

prudhvid commented on vision_stack/buoy_detect/region_growing/src/grow.cpp:
```diff
@@ -0,0 +1,290 @@
+        stringstream ssy(s.substr(pos + 1));
+        ssx >> x;
+        ssy >> y;
+
```
These contain lots of repetitions, change these to lambda functions in C++11; If you're using an old compiler in CMakeLists make sure you use C++11

Modify the below code to this

``` C++
    auto bfsHelper = [&](int x1, int y1)
    {
        if (colorDistance(seed, input.at<Vec3b>(x1, y1), colorThreshold, whiteThreshold))
        {
            reach[x1][y1] = true;
            posQueue.push_back(make_pair(x1, y1));
            modifyPixel(filled, x1, y1, colorflag);
            ++count;
        }
        else
        {
            modifyPixel(edgeMap, x1, y1, colorflag);
        }
    }

    if (x + 1 < input.rows && (!reach[x + 1][y]))
    {
        bfsHelper(x + 1, y);
    }
    if (y + 1 < input.cols && (!reach[x][y + 1]))
    {
        bfsHelper(x, y + 1);
    }
//soo on
```

pranaypratyush replied:
@Nightfury18 learn about auto (std::future) and lambda function

pranaypratyush commented:
@prudhvid couldn't find a single flaw in MY code. LOL. I must be pro.

pranaypratyush committed:
'predict_block' function from 'vw_detect' was made public

prudhvid commented on vision_stack/buoy_detect/vw_detect/src/vw_detect.h:
```diff
@@ -0,0 +1,57 @@
+{
+public:
+    int n_threads;
+    uint_least8_t hash_table[256][256][256];
```
@pranaypratyush  :+1:  awesome work on types

pranaypratyush replied:
what can i say. I am the pro.

prudhvid commented on vision_stack/buoy_detect/vw_detect/src/vw_detect.cpp:
```diff
@@ -0,0 +1,143 @@
+using namespace std;
+using namespace cv;
+
+void vw_detect::vw_detect_init(char *path_to_hash)
```
Also make it a habit to have const wherever possible.

prudhvid commented on vision_stack/buoy_detect/vw_detect/src/vw_detect.cpp:
```diff
@@ -0,0 +1,143 @@
+    Mat hsv_image;
+    cvtColor(original, hsv_image, CV_BGR2HSV);
+    //    prediction = Mat(original.rows, original.cols, CV_8UC3, Scalar(0, 0, 0));
+    for (int i = 0; i <= 255; i += original.rows / n_threads)
```
I'm not sure but Is it i <= 255 or i<255?

pranaypratyush replied:
it's correct.

prudhvid replied:
Yeah cool. I thought its 256 😜

pranaypratyush committed:
Major bug fixes
+ still getting exceptions
+ gpu accelerated optimizations incoming

pranaypratyush committed:
Adds the 'chrono' time measurement commands

pranaypratyush committed:
Bug fixes, trying new algo
+ Added the 'build_hash.sh' script

icyflame commented on vision_stack/buoy_detect/src/integration.cpp:
```diff
@@ -116,8 +125,14 @@ void imageCallback(const sensor_msgs::ImageConstPtr& msg)
     cv_ptr->image = final_image;
     final_pub->publish(cv_ptr->toImageMsg());
+    auto t4 = clk::now();
+    cout << "time re receive and queue : " << duration_cast<milliseconds>(t2 - t1).count() <<endl;
```
@pranaypratyush @Nightfury18 Please use `ROS_DEBUG` and other logging functions that are provided by ROS, that makes setting log levels easy. It's hard to incorporate them once the code is in the master branch.

pranaypratyush replied:
sure. i think `ROS_DEBUG` sounds like the perfect severity level for this.

icyflame replied:
@pranaypratyush Just follow the normal severity rules. Anything that doesn't hamper information during an actual run will go into `DEBUG`. Information required all the time will go into `INFO`, and errors will go into `ERROR`. ROS handles coloring of messages and tagging with timestamps too, so, makes debugging a lot easier when testing in the pool.

icyflame commented on vision_stack/buoy_detect/README.md:
```diff
@@ -0,0 +1,24 @@
+BUOY_DETECT
+-----------
+This repository contains the codes that uses ML (Machine Learning) techniques to detect the buoy. This is a combination of 3 parts - 
+<br>
```
@pranaypratyush Don't use `<br>` inside markdown. Rather, use two new lines. (You can also add two whitespaces on the previous line)

pranaypratyush replied:
@athityakumar your code.

pranaypratyush replied:
why can't I tag him ?

athityakumar replied:
I guess you've tagged me @pranaypratyush :wink: 

Sure, @icyflame - that'll be taken care of, from next time. :smile:

icyflame replied:
@athithyakumar : Fix the problems in this file as well.

@pranaypratyush You can't tag him because he isn't in the organisation yet. I will add him now.

athityakumar replied:
@icyflame : Fixed it.

@pranaypratyush : Sent you a PR at https://github.com/pranaypratyush/kraken_3.0/pull/2

athityakumar committed:
Added more documentation

(1) Included steps to be followed to run the scripts with & without ROS, for benifit of future new-recruits.
(2) Made the changes, suggested by @icyflame

pranaypratyush committed:
Merge pull request #2 from athityakumar/buoy_detect

Added more documentation

pranaypratyush committed:
Major bug fixes
+ Debugged the 'build_hash.sh' script
+ relevant changes were made to 'vw_detect' package

pranaypratyush committed:
Merge branch 'buoy_detect' of https://github.com/pranaypratyush/kraken_3.0 into buoy_detect

pranaypratyush commented:
@icyflame This branch would remain unmaintained for some time, I would be pushing the "jugadu" code on a new branch. Is that okay ?

icyflame commented:
@pranaypratyush Sure, I am closing this PR for now. Keep this in the pipeline, and get back to it whenever we are done with the basic Buoy task.

@prudhvid Thanks for your time spent reviewing this. This is **not** being shelved. For now, @pranaypratyush and @Nightfury18 believe that a _simple, ML-based_ approach should be able to solve the Buoy Task problem, which is what we need to demonstrate to Prof. CSK. We will eventually get back to this PR.
